Plugin system improvements

default-writer opened the pull request:
I did some improvements of plugin system

- added interfaces, enchanting plugin system
- added HubOptions section in appsettings.json
- added RegisterPluginServices to register and load themes and generic plugins

csharpfritz commented:
You've added a number of interfaces here that I'm not sure I agree with 

Why add an IPluginManifest? PluginManifest is a POCO and can live in the Plugins project. Plugins themselves don't need access to this representation 

Why make ApplicationState an interface? This currently is a singleton in the web project.

default-writer commented:
Ok, I will explain my point of view:

##  ApplicationState 

```cs
public class ApplicationState : IApplicationState
{
	public long? MaximumUploadSizeMB { get; set; } = 10;
	public CurrentThemeRecord? CurrentTheme { get; set; } = null;
}
```

Class originally had responsibilities including 

- AddPlugin (plugin management)
- SetTheme (theme management)
- Load/Save to disk (persistence)

I decided that is not very correct to have this class to save and i just renamed `ApplicationState` to `ApplicationStateManager` which does the same as ApplicationState and less, i.e. i moved `public Dictionary<string, PluginManifest> Plugins` into another place.

Since IApplicationStateManager is a singleton, and it holds ApplicationState, it can be okay to run it.


For some reason I decided to add an interface definition for all classes in SharpSite.Plugins library, and more.

I used `IApplicationState` interface for the reason that it can be accessible using interface `IApplicationStateManager`

wich is also implements 'IStartupManager' interface

```cs

public interface IApplicationStateManager: IStartupManager
{
	IReadOnlyDictionary<string, IPluginManifest> Themes { get; }

	IApplicationState CurrentState { get; }

	public Type? ThemeType { get; }

	Task SaveState();

	void SetTheme(IPluginManifest manifest);
}
```

## IPluginManifest

`IPluginManifest` is used in `IPluginAssembly`

```cs
public interface IPluginAssembly
{
	IPluginManifest Manifest { get; }

	void LoadContext();
	void UnloadContext();
}
```

It allows us or plugins developers or core developers to implement their own vaildation check or create their own implementations of `PluginAssemblyManager` or even use different version of the same library.

It works well with this POCO class and we can get all data and custom logic without having to know real implementation in SharpSite.Web project. So we are breaking dependency and make this settings visible to others. We can comment out some interfaces to hide some internals from public.

default-writer commented:
I just want to hilight that I am using 3 singleton objects 


- PluginAssemblyManager
- PluginManager
- ApplicationStateManager (holds ApplicationState POCO with 2 options for a moment)

and register them the same way as Postgres plugins (you mentioned this as an option on your stream recently)

```cs
public class RegisterPluginServices : IRegisterServices, IRunAtStartup
{
	public IHostApplicationBuilder RegisterServices(IHostApplicationBuilder builder, bool disableRetry = false)
	{
		builder.Services.AddSingleton<IPluginAssemblyManager, PluginAssemblyManager>();
		builder.Services.AddSingleton<IPluginManager, PluginManager>();
		builder.Services.AddSingleton<IApplicationStateManager, ApplicationStateManager>();

		return builder;
	}

	public async Task RunAtStartup(IServiceProvider services)
	{
		PluginManager.Initialize();

		var pluginManager = services.GetRequiredService<IPluginManager>();
		await pluginManager.LoadAtStartup();

		var themeManager = services.GetRequiredService<IApplicationStateManager>();
		await themeManager.LoadAtStartup();
	}
}
```
